File: Graphs/RRT.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation

def intersect(p1, p2, p3, p4):
    """
    return True if lines intersect
    """

    x1, y1 = p1 # a point on the first line
    x2, y2 = p2 # another point on the first line
    x3, y3 = p3 # a point on the second line
    x4, y4 = p4 # another point on the second line

    #check paralell:
    if (x1-x2)*(y3-y4) == (y1-y2)*(x3-x4):
        return False

    px = ((x1*y2-y1*x2)*(x3-x4)-(x1-x2)*(x3*y4-y3*x4))/((x1-x2)*(y3-y4)-(y1-y2)*(x3-x4))
    py = ((x1*y2-y1*x2)*(y3-y4)-(x1-x2)*(x3*y4-y3*x4))/((x1-x2)*(y3-y4)-(y1-y2)*(x3-x4))

    #scalar value desciding point of intersection on line 1
    t = ((x1-x3)*(y3-y4)-(y1-y3)*(x3-x4))/((x1-x2)*(y3-y4)-(y1-y2)*(x3-x4)) #scalar value desciding point of intersection on line 1

    #scalar value desciding point of intersection on line 2
    u = ((x1-x3)*(y1-y2)-(y1-y3)*(x1-x2))/((x1-x2)*(y3-y4)-(y1-y2)*(x3-x4))

    _ = [t, u]

    if all(0<= i <= 1 for i in _):
        #print('points = {}'.format(((x1, y1),(x2, y2), (x3, y3), (x4, y4))))
        return True

    return False


def RRT(num_points, connectivity, starting_point, seed):
    """ Generates random points using the a version of the Rapid Random Trees algorithmself.
    low connectivity means high spread and vice versa.
    """
    np.random.seed(seed)
    N = num_points
    a = np.multiply(np.ones((N, 2)), starting_point)
    points = np.zeros((N, 2))
    points[0] = starting_point
    for i in range(1, N):
        b = (connectivity*num_points)*np.random.uniform(low = 0, high = 1, size = (1, 2))
        w = np.argmin(np.sum((a-b)**2, 1))
        n = a[w, :]
        k = np.arctan2(b[0][1]-n[1], b[0][0]-n[0])
        c = np.array([n[0]+np.cos(k), n[1]+np.sin(k)])
        a[i, :] = c
        points[i] = (c[0]+n[0])/2, (c[1]+n[1])/2,

    a = points[:, 0]/np.max(points[:, 0])
    b = points[:, 1]/np.max(points[:, 1])
    return np.column_stack((a,b))



def connect_RRT(points, max_norm):
    connections = [] #store connections [(point1, point1), ..]
    N = len(points[:, 0])
    for i in range(1, N): #skip starting point
        iter = 0
        for j in range(N):
            if i > j: #only connect from lower to higher index
                if np.linalg.norm(points[i, :]-points[j, :]) < max_norm:
                    iter += 1
                    if iter < 2: #restrict connections in clusters, only 1 brancing allowed
                        connections.append((j, i))

    # #exclude nodes with intersect connections
    intersections = []
    for i in range(len(connections)):
        for j in range(len(connections)):
            if i<j:
                if connections[i][0] not in connections[j] and connections[i][1] not in connections[j]:
                    if intersect(points[connections[i][0], :], points[connections[i][1], :], \
                        points[connections[j][0], :], points[connections[j][1], :]):
                        intersections.append(j)#print intersect connection

    #exclude nodes without connection
    lines = np.array(connections)
    lines = np.delete(lines, intersections, axis = 0)
    points_removed = []
    i = 1
    while i < len(points):
        if (i in lines[:, 1]) == False:
            points[i, :] = None #np.delete(points, i, 0)
            points_removed.append(i)
            if np.where(lines[:, 0]==i)[0].size > 0:
                lines = np.delete(lines, np.where(lines[:, 0]==i)[0], axis = 0)
        i += 1

    print("intersections: {}, points removed: {}, seed: {}".format(intersections, points_removed, seed))
    return points, lines


def plot_graph(nodes, lines, view = True, annotate = False, save = False):
    fig, ax = plt.subplots()
    ax.scatter(nodes[:, 0], nodes[:, 1], s = 0.1)
    n = range(len(points[:, 0]))
    for i, j in lines:
        ax.plot([nodes[i, 0], nodes[j, 0]], [nodes[i, 1], nodes[j, 1]], 'k-', lw = 0.1)

    if annotate == True:
        for i, txt in enumerate(n):
            ax.annotate(txt, (nodes[i, 0], nodes[i, 1]),  fontsize = 6)
    if save:
        plt.savefig('graph_image.jpg', dpi = 30)
    if view:
        plt.show()


def create_animation(points):
    fig = plt.figure("RRT")
    ax = fig.add_subplot(111, aspect='equal',
                            xlim=(0, np.max(points[:, 0])), ylim=(0, np.max(points[:, 1])))
    p, = plt.plot(points[0, 0], points[0, 1], 'ko')
    line, = plt.plot([points[0, 0], points[0, 1]], 'k--')

    def animate(i):
        p.set_data(points[i, 0], points[i, 1])
        line.set_data(points[0:i, 0], points[0:i, 1])
        return p, line

    anim = animation.FuncAnimation(fig, animate,  frames=25, \
                                           interval=25, blit=True)
    anim.save('rrt.mp4', fps = 1)

if __name__ == "__main__":
    for seed in range(1, 15):
        starting_point = np.array([0, 0])
        points = RRT(35, 1, starting_point, seed)
        nodes, lines = connect_RRT(points, 0.1)
        plot_graph(nodes, lines, view = True, annotate = True)

File: Graphs/RRT.jl
using Distributions
using Random
using Plots
using LinearAlgebra:norm

function RRT(num_points, connectivity, seed; starting_point=[0.0,0.0])
    Random.seed!(seed)
    N = num_points
    x0 = starting_point[1]
    y0 = starting_point[2]
    points = zeros(N,2)
    points[1, :] = [x0, y0]
    a = hcat([x0 for i in 1:N], [x0 for i in 1:N])
    for i in 2:N
        b = (connectivity*N)*rand(Uniform(0, 1), 1,2)
        w = sum((a.-b).^2, dims = 2)
        idx = map(argmin, eachcol(w))[1]
        n = a[idx, :]
        k = atan(b[2]-n[2], b[1]-n[1])
        c = [n[1]+cos(k), n[2]+sin(k)]
        a[i, :] = c
        points[i, :] = [(c[1]+n[1])/2, (c[2]+n[2])/2]
    end
return points
end


function line_intersection(p1, p2, p3, p4)
    """
    return true and point of intersection if lines intersect
    """

    x1, y1 = p1 # a point on the first line
    x2, y2 = p2 # another point on the first line
    x3, y3 = p3 # a point on the second line
    x4, y4 = p4 # another point on the second line

    #check paralell:
    if (x1-x2)*(y3-y4) == (y1-y2)*(x3-x4)
        return false
    end

    px = ((x1*y2-y1*x2)*(x3-x4)-(x1-x2)*(x3*y4-y3*x4))/((x1-x2)*(y3-y4)-(y1-y2)*(x3-x4))
    py = ((x1*y2-y1*x2)*(y3-y4)-(y1-y2)*(x3*y4-y3*x4))/((x1-x2)*(y3-y4)-(y1-y2)*(x3-x4))

    #scalar value desciding point of intersection on line 1
    t = ((x1-x3)*(y3-y4)-(y1-y3)*(x3-x4))/((x1-x2)*(y3-y4)-(y1-y2)*(x3-x4)) #scalar value desciding point of intersection on line 1

    #scalar value desciding point of intersection on line 2
    u = ((x1-x3)*(y1-y2)-(y1-y3)*(x1-x2))/((x1-x2)*(y3-y4)-(y1-y2)*(x3-x4))

    s = [t, u]

    if all(>=(0), s) & all(<=(1), s)
        return true, [px,py]
    end

return false, []
end



function connect_RRT(points, ;max_norm::Float64 = 1.0)
    connections = [] #store connections [(point1, point1), ..]
    POI = [] #if intersect, append here
    nodes = points
    N = size(points[:, 1])[1]
    for i in 2:N #skip starting point
        iter = 0
        for j in 1:N
            if norm(hcat(points[i, :]-points[j, :])) < max_norm
                iter += 1
                if iter < 2
                    push!(connections, (j, i))
                end
            end
        end
    end
    #check if intersect
    for i in 1:length(connections)
        for j in 1:length(connections)
            if i<j
                if (connections[i][1] ∉ connections[j]) & (connections[i][2] ∉ connections[j])
                    bool, p = line_intersection(points[connections[i][1], :], points[connections[i][2], :],
                        points[connections[j][1], :], points[connections[j][2], :])
                    if bool
                        push!(POI, p)
                        println("intersection at: ", POI)
                    end
                end
            end
        end
    end
    nodes = nodes/maximum(nodes) #normalize
return nodes, hcat(collect.(connections)...)', POI
end

false && begin
    #visual stress test
    for seed in rand(1:1000, 100)
        println("seed: ", seed)
        global points = RRT(30, 1, seed)
        nodes, lines, POI = connect_RRT(points)

        scatter(nodes[:, 1], nodes[:, 2])

        for (i,j) in eachrow(lines)
            plot!([nodes[i, 1], nodes[j, 1]],[[nodes[i, 2], nodes[j, 2]]], color = :blue, show = true, legend = false,
            xlim = (0,1.01), ylim = (0,1.01))
        end
        sleep(1.0) #time delay between plots
    end
    #Figure
    function create_figure(seeds)
        points = RRT(30, 1, seeds[1])
        nodes, lines, POI = connect_RRT(points)
        a = scatter(nodes[:, 1], nodes[:, 2])
        for (i,j) in eachrow(lines)
            plot!([nodes[i, 1], nodes[j, 1]],[[nodes[i, 2], nodes[j, 2]]], color = :blue, legend = false)
        end
        points = RRT(30, 1, seeds[2])
        nodes, lines, POI = connect_RRT(points)
        b = scatter(nodes[:, 1], nodes[:, 2])
        for (i,j) in eachrow(lines)
            plot!([nodes[i, 1], nodes[j, 1]],[[nodes[i, 2], nodes[j, 2]]], color = :blue, legend = false)
        end
        points = RRT(30, 1, seeds[3])
        nodes, lines, POI = connect_RRT(points)
        c = scatter(nodes[:, 1], nodes[:, 2])
        for (i,j) in eachrow(lines)
            plot!([nodes[i, 1], nodes[j, 1]],[[nodes[i, 2], nodes[j, 2]]], color = :blue, legend = false)
        end
        points = RRT(30, 1, seeds[4])
        nodes, lines, POI = connect_RRT(points)
        d = scatter(nodes[:, 1], nodes[:, 2])
        for (i,j) in eachrow(lines)
            plot!([nodes[i, 1], nodes[j, 1]],[[nodes[i, 2], nodes[j, 2]]], color = :blue, legend = false)
        end
        e = plot(a, b, c, d, layout=(1,4), show = true, size =(1200, 300))
        savefig(e, "4_graphs")
    end
    #create_figure([1,2,3,4])
end
